File: 33-search-in-rotated-sorted-array/33-search-in-rotated-sorted-array.java
class Solution {
    public int search(int[] nums, int target) {
        int start = 0;
        int end = nums.length - 1;
        
        while(start <= end) {
            int mid = (start + end) / 2;
            
            if (nums[mid] == target) {
                return mid;
            }
            if (nums[start] == target) {
                return start;
            }
            if (nums[end] == target) {
                return end;
            }
            
            int s = nums[start];
            int m = nums[mid];
            int e = nums[end];
            
            if (m < e) {
                if (m < target && target < e) start = mid + 1;
                else end = mid - 1;
            } else {
                if (s < target && target < m) end = mid - 1;
                else start = mid + 1;
            }
        }
        
        return -1;
    }
}

File: 33-search-in-rotated-sorted-array/33-search-in-rotated-sorted-array.scala
object Solution {
    def search(nums: Array[Int], target: Int): Int = {
        var s = 0
        var e = nums.length - 1
        var res = -1
        
        while(res == -1 && s <= e) {
            val mid = (s + e) / 2
            val v = nums(mid)
            val vs = nums(s)
            val ve = nums(e)
            
            if (v == target) {
                res = mid 
            } else if (
                (vs < v && target >= vs && target <= v) ||
                (vs > v && (target >= vs || target <= v))
            ) { // the possible answer is on the left
                e = mid - 1
            } else { // the possible answer is on the right
                s = mid + 1
            }
        }
        
        res
    }
}